src/neo/SmartContract/Native/Tokens/GasToken.cs

```diff
@@ -32,11 +32,55 @@ internal override bool Initialize(ApplicationEngine engine)
         protected override bool OnPersist(ApplicationEngine engine)
         {
             if (!base.OnPersist(engine)) return false;
+
+            long oracleFee = 0;
+            long networkFee = 0;
+
             foreach (Transaction tx in engine.Snapshot.PersistingBlock.Transactions)
+            {
+                networkFee += tx.NetworkFee;
                 Burn(engine, tx.Sender, tx.SystemFee + tx.NetworkFee);
+
+                switch (tx.Version)
+                {
+                    case TransactionVersion.OracleRequest:
+                    case TransactionVersion.OracleResponse:
+                        {
+                            oracleFee += tx.SystemFee;
+                            break;
+                        }
+                }
+            }
+
             ECPoint[] validators = NEO.GetNextBlockValidators(engine.Snapshot);
             UInt160 primary = Contract.CreateSignatureRedeemScript(validators[engine.Snapshot.PersistingBlock.ConsensusData.PrimaryIndex]).ToScriptHash();
-            Mint(engine, primary, engine.Snapshot.PersistingBlock.Transactions.Sum(p => p.NetworkFee));
+            Mint(engine, primary, networkFee);
+
+            if (oracleFee > 0)
+            {
+                primary = Oracle.GetOracleMultiSigAddress(engine.Snapshot);
+
+                if (BalanceOf(engine.Snapshot, primary) < 10_0000_0000)
+                {
+                    // In order to have funds for make transfers we need to ensure
+                    // that the oracle multi signature have enough gas
+
+                    Mint(engine, primary, oracleFee);
+                    return true;
+                }
+
+                // Distribute oracle fee
+
+                validators = Oracle.GetOracleValidators(engine.Snapshot);
+                var amount = oracleFee / validators.Length;
+
+                foreach (var validator in validators)
+                {
+                    // Transfer check the oracle's multi signature
```

Comment:
Will it be better if we could move this incentive model to `OracleContract.OnPersist` method?

Reply:
Then Gas will burn all of them first

Reply:
Maybe it's okay to put here, as it's  more efficient.